lib/nrf_modem_lib/nrf91_net_l2/nrf91_net_l2.c

```diff
@@ -0,0 +1,328 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ *
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/types.h>
+#include <zephyr/net/offloaded_netdev.h>
+#include <zephyr/net/net_l2_connectivity.h>
+#include <zephyr/logging/log.h>
+#include <modem/nrf_modem_lib.h>
+#include <modem/lte_lc.h>
+#include <modem/pdn.h>
+
+#include "ip_addr_helper.h"
+
+LOG_MODULE_REGISTER(nrf91_net_l2, CONFIG_NRF91_NET_L2_LOG_LEVEL);
+
+/* This library don't support the CONFIG_NRF_MODEM_LIB_SYS_INIT option.
+ *
+ * If this option is set, initialization of the nRF Modem library occurs
+ * automatically, at the same time as the Zephyr network stack initializes the nRF91 NET L2 layer.
+ * At SYS INIT - POST_KERNEL init level.
+ *
+ * Due to this we don't support the aforementioned option when building with this library.
+ * We need to control the initialization order to prevent a potential race in case SYS INIT level
+ * and priority changes.
+ */
+BUILD_ASSERT(!IS_ENABLED(CONFIG_NRF_MODEM_LIB_SYS_INIT),
+	     "System initialization of nRF Modem library is not supported");
```

Comment:
Its definitely doable. I just wonder if its worth the extra complexity. Is this option really something that we want to recommend to users? At least right now its very clear the sequence that needs to be followed when (re)init the modem.
Something that definitely won't be the case if supporting option.

@lemrey I'm curious what you take is here.

Reply:
In my opinion `SYS_INIT` should be used for simple runtime initialization tasks that can't fail. The reason for that is if any error happens during `SYS_INIT`, there is no way to return that error to the application, unless it is stored somewhere.

In case of `nrf_modem_lib_init()`, any failures, e.g. FOTA, low power, are not reported to the application, unless the application manually setups a callback to receive/store that error. I do not see any advantages to this approach compared to letting the application call `nrf_modem_lib_init` by itself.

On the contrary this gives some headaches. Inherently, it is less straightforward, but more importantly it introduces two ways to do the same thing, i.e. turn on the modem. This in turn adds complexity to samples and libraries which want/have to support both use-cases (auto-init and no auto-init).

There's also the fact that turning on the modem can take _minutes_ when the modem firmware is being updated. So the application will be stuck in `SYS_INIT` for over a minute after a reset, before it gets the chance to run to `main()`.

All in all, I do not think that there are any benefits to using `NRF_MODEM_LIB_SYS_INIT`. It is sparing the application  _one_ function call but adds far more complexity and cost elsewhere. Keep in mind that other components that currently rely on `NRF_MODEM_LIB_SYS_INIT` to also initialize themselves during `SYS_INIT`, could simply setup a hook on libmodem initialization and initialize automatically whenever the modem is turned on instead.

My conclusion is that `NRF_MODEM_LIB_SYS_INIT` should be deprecated and we should move away from it.

Reply:
Generally, I agree with @lemrey that `NRF_MODEM_LIB_SYS_INIT` should be deprecated. As you point out, it creates quite a few headaches, and it would be nice not to have to account for that in our libraries.

I was under the false impression that a lot of our libraries require `NRF_MODEM_LIB_SYS_INIT`, but looking closer, only the AT_HOST library requires it, and I don't think it would be too hard to remove that dependency in another PR, so this isn't nearly as big of a deal as I thought. I'll let the issue rest, I think you both are correct